---
title: "Quality Control of AnVILHMMRATAC Data"
author:
- name: Kayla Interdonato
  affiliation: Roswell Park Comprehensive Cancer Center, Buffalo, NY
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        toc: true
        toc_float: true
vignette: >
  %\VignetteIndexEntry{A_QC}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

The goal of the `AnVILHMMRATAC` package is to provide users all the tools to
perform a `HMMRATAC` analysis utilizing AnVIL. This package contains all the
files (except for raw reference genome and chromosome information) needed to
complete the analysis.

This vignette will demonstrate to the user how to perform a quality check of
the alignment using the `r Biocpkg("ATACseqQC")` package. The bam files for each
file will be obtained from the Google bucket of this project.

# Installation

The following code will install the latest version of the `AnVILHMMRATAC`
package from Bioconductor.

```{r install, eval = FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("AnVILHMMRATAC")
```

Then to load `AnVILHMMRATAC`:

```{r load, message = FALSE}
library(AnVILHMMRATAC)
```

We will also need the `AnVIL` package for certain functionality so we will
install and load it now.

```{r anvil}
pkgs = c("Bioconductor/AnVIL", "Rsamtools", "ATACseqQC",
    "BSgenome.Hsapiens.NCBI.GRCh38", "MotifDb") 
BiocManager::install(pkgs)

suppressPackageStartupMessages({
    library(AnVIL)
    library(Rsamtools)
    library(ATACseqQC)
    library(BSgenome.Hsapiens.NCBI.GRCh38)
    library(MotifDb)
})
```

# Locate the files

We will be performing QC on each of the bam files produced from the 
AnVILHMMRATAC workflow. In order to run this QC we must first locate the bam 
files. We have saved the sorted bam files from the workflow, so sorting is not
needed. If your bam files have not been sorted, we suggest doing this first.

As part of the workflow, we saved the default outputs. This means that the paths
to the outputs were saved in the corresponding entry in the 'participant_set'
table. We will obtain the paths from here.

```{r participant_set}
avworkspace("bioconductor-rpci-anvil/HMMRATAC")
part_set <- avtable("participant_set") %>% 
    filter(name == "AnVILHMMRATAC_2021-05-26T15-32-59") %>%
    tidyr::unnest()
```

Now that we have the participant set table, with the locations for the outputs,
we can pull out the needed files. For the first QC step we will need to copy the
sorted bam and sorted bam bai files copied to the working directory.

```{r copy_files}
sorted_bams <- part_set %>% pull("sorted_bam.items")
AnVIL::gsutil_cp(sorted_bams, getwd())

sorted_bam_bais <- part_set %>% pull("sorted_bam_bai.items")
AnVIL::gsutil_cp(sorted_bam_bais, getwd())
```

# Perform QC

Next we will use the `ATACseqQC` package to assess the quality of the
alignment. We demonstrate this step on one sample, but will be applicable to any
of the samples.

```{r libraryComplexity}
fls <- list.files(".", pattern = "*.bam$")

estimateLibComplexity(readsDupFreq(fls[1])) ## this took about 9 min. to run
```

```{r fragSizeDist}
fragSize <- fragSizeDist(fls[1], "Frag Sizes")
```

We can also perform footprint identification with `ATACseqQC`.

```{r footprintID}
CTCF <- query(MotifDb, c("CTCF"))
CTCF <- as.list(CTCF)
seqlev <- "1"

sigs <- factorFootprints(fls[1],
    pfm = CTCF[[1]],
    genome = Hsapiens,
    min.score = "90%",
    seqlev = seqlev,
    upstream = 100,
    downstream = 100)

featureAlignedHeatmap(sigs$signal,
    feature.gr = reCenterPeaks(sigs$bindingSites,
        width = 200+width(sigs$bindingSites[1])),
    annoMcols = "score",
    sortBy = "score",
    n.tile = ncol(sigs$signal[[1]]))
```                                                 

# Session Information

```{r}
sessionInfo()
```
