---
title: "Downstream Analysis of AnVILHMMRATAC Data"
author:
- name: Kayla Interdonato
  affiliation: Roswell Park Comprehensive Cancer Center, Buffalo, NY
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        toc: true
        toc_flow: true
vignette: >
  %\VignetteIndexEntry{B_Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

This vignette will walk a user through how to run a full downstream analysis of
HMMRATAC on the output data from the `AnVILHMMRATAC` workflow. The output data
will be retrieved and utilized in functions from the `r Biocpkg("DiffBind")`,
the `r Biocpkg("rGREAT")`, and the `r Biocpkg("ChIPseeker")` packages. 

# Installation

```{r setup}
library(AnVILHMMRATAC)
```

```{r install}
pkgs = c("Bioconductor/AnVIL", "DiffBind", "rGREAT", "ChIPseeker", 
    "TxDb.Hsapiens.UCSC.hg19.knownGene", "org.Hs.eg.db")
BiocManager::install(pkgs)

suppressPackageStartupMessages({
    library(AnVIL)
    library(DiffBind)
    library(rGREAT)
    library(ChIPseeker)
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    library(org.Hs.eg.db)
})
```

# Obtaining the data

The location of the files needed for the following analyses can be obtained from
the 'participant_set' table. We will create a table of these locations so that
we can copy them to the workspace when needed.

```{r part_set}
avworkspace("bioconductor-rpci-anvil/HMMRATAC")
part_set <- avtable("participant_set") %>%
    filter(name == "AnVILHMMRATAC_2021-05-26T15-32-59") %>%
    tidyr::unnest()
```

The sorted bam files should already be copied to the current working directory
from the previous vignette. We will also need the summit bed files copied to
the working directory for the DiffBind analysis.

```{r summit_beds}
summit_beds <- part_set %>% pull("summits_bed.items")
AnVIL::gsutil_cp(summit_beds, getwd())
``` 

# DiffBind

The DiffBind package specializes in identifying sites that are differentially
bound between two sample groups. It also contains functionality for processing
peak sets, counting sequencing reads overlapping intervals in peak sets, and
identifying statistically significant differentially bound sites based on
evidence of binding affinity.

Here, we will use metadata from the 'participant' table since this table
discribes the treatments from the study. Then we will perform count, contrast,
analyze and report from DiffBind.

```{r samplesheet}
part_tbl <- avtable("participant")
sub_tbl <- as.data.frame(part_tbl[7:12, -c(2:4)])
colnames(sub_tbl) <- c("SampleID", "Tissue", "Replicate", "Condition")

sorted_bams <- file.path(getwd(), "SRR8912%s.sorted.bam")
sub_tbl$bamReads <- sprintf(sorted_bams, 75:80)

peaks <- file.path(getwd(), "SRR8912%s_summits.bed")
sub_tbl$Peaks <- sprintf(peaks, 75:80)

sub_tbl$PeakCaller <- rep("bed", 6)
```

```{r diffbind}
dba <- dba(sampleSheet = sub_tbl)
count <- dba.count(dba, bParallel = FALSE)
contrast <- dba.contrast(count, contrast = c("Condition", "Day1", "Day3"))
analyze <- dba.analyze(contrast)
bed <- dba.report(analyze)
```

# rGREAT

`rGREAT` is a package that can read peak files and infer which pathways are
enriched using functional enrichemnt analysis.

```{r rGREAT}
## probably won't work because chromosome names in bed
## don't have 'chr' prefix
job <- submitGreatJob(bed)
tb <- getEnrichmentTables(job)
look <- tb[[1]]
res <- look[look$Hyper_Adjp_BH < 0.05,]
```

# ChIPseeker

The ChIPseeker package is meant to perform statistical testing of significant
overlap among ChIP seq datasets. It also incorporates open access database GEO
to allow users to compare their dataset to those available on GEO.

First we will need to copy over the gappedPeak files from the HMMRATAC output to
our current working directory.

```{r gpCopy}
gP_paths <- part_set %>% pull("filtered_gappedPeak.items")
AnVIL::gsutil_cp(gP_paths, getwd())
```

```{r chipseeker}
fls <- list.files(".", pattern = "*.filteredPeaks.gappedPeak$")
peak <- readPeakFile(fls[8])
```

Display peak locations across entire genome.

```{r covplot}
covplot(peak, weightCol = "X33.0") ## I'm not sure if this is the correct column
```

Perform peak annotation.

```{r peakAnnotation}
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
peakAnno <- annotatePeak(fls[8],
    tssRegion = c(-3000, 3000),
    TxDb = txdb,
    annoDb = "org.Hs.eg.db")

plotAnnoPie(peakAnno)

upsetplot(peakAnno, vennpie = TRUE)

plotDistToTSS(peakAnno,
    title = "Distribution of transcription factor-binding loci relative to
TSS")
```

# Session Information

```{r}
sessionInfo()
```
